### ISSUE TITLE
Templates doubling

### ISSUE THREAD

**Issue by victorvoronin** [openshift/jenkins-sync-plugin#197]
I have been added configmap with 2 templates. As a result first template added once in Jenkins config and second - twice. Then I disable Sync plugin, and enable again. Now I see 6 copies of the first template and 2 copies on second. How can I catch this unpredictable issue?

**Comment by gabemontero** [comment]
Because of how the k8s plugins stores pod templates, there has been some difficulty with avoiding duplicates in the past

At least one fix went in for this about 5 months ago with commit 753907b376d9cc4991a4bfad5f86d91cd46644e6 though that should be in v0.1.28

If you are using say a v3.6 version of the openshift jenkins image, you might not have this fix unless you explicitly upgraded the plugin

@victorvoronin - could you clarify the version of the sync plugin, and perhaps the openshift jenkins image if you use that, that you are currently using?

**Comment by victorvoronin** [comment]
Our configuration:
Openshift origin 3.7
Jenkins 2.89.3
Openshift Sync Plugin 1.0.2

**Comment by gabemontero** [comment]
should have the fixes in this space ... I'm trying to reproduce locally ...

**Comment by gabemontero** [comment]
Yeah I'm not having any luck reproducing this .. multiple templates, multiple config maps, restarting the plugin.  That said, I wouldn't be surprised if there is some timing window between watches and relist that is hitting you but not me, and coupled with the non-optimal API the k8s plugin gives us, voila.

Let's go about these steps @victorvoronin to see if we can sort it out for you:

1) First, a sanity check on the config map contents; I assume you have something that looks like:

```
apiVersion: v1
items:
- apiVersion: v1
  data:
    jenkins-slave: |-
      <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
        <inheritFrom></inheritFrom>
        <name>jenkins-slave</name>
        <instanceCap>2147483647</instanceCap>
        <idleMinutes>0</idleMinutes>
        <label>jenkins-slave</label>
        <serviceAccount>jenkins</serviceAccount>
        <nodeSelector></nodeSelector>
        <volumes/>
        <containers>
          <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
            <name>jnlp</name>
            <image>openshift/jenkins-slave-maven-centos7</image>
            <privileged>false</privileged>
            <alwaysPullImage>false</alwaysPullImage>
            <workingDir>/tmp</workingDir>
            <command></command>
            <args>${computer.jnlpmac} ${computer.name}</args>
            <ttyEnabled>false</ttyEnabled>
            <resourceRequestCpu></resourceRequestCpu>
            <resourceRequestMemory></resourceRequestMemory>
            <resourceLimitCpu></resourceLimitCpu>
            <resourceLimitMemory></resourceLimitMemory>
            <envVars/>
          </org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
        </containers>
        <envVars/>
        <annotations/>
        <imagePullSecrets/>
        <nodeProperties/>
      </org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
    jenkins-slave-2: |-
      <org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
        <inheritFrom></inheritFrom>
        <name>jenkins-slave-2</name>
        <instanceCap>2147483647</instanceCap>
        <idleMinutes>0</idleMinutes>
        <label>jenkins-slave</label>
        <serviceAccount>jenkins</serviceAccount>
        <nodeSelector></nodeSelector>
        <volumes/>
        <containers>
          <org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
            <name>jnlp</name>
            <image>openshift/jenkins-slave-maven-centos7</image>
            <privileged>false</privileged>
            <alwaysPullImage>false</alwaysPullImage>
            <workingDir>/tmp</workingDir>
            <command></command>
            <args>${computer.jnlpmac} ${computer.name}</args>
            <ttyEnabled>false</ttyEnabled>
            <resourceRequestCpu></resourceRequestCpu>
            <resourceRequestMemory></resourceRequestMemory>
            <resourceLimitCpu></resourceLimitCpu>
            <resourceLimitMemory></resourceLimitMemory>
            <envVars/>
          </org.csanchez.jenkins.plugins.kubernetes.ContainerTemplate>
        </containers>
        <envVars/>
        <annotations/>
        <imagePullSecrets/>
        <nodeProperties/>
      </org.csanchez.jenkins.plugins.kubernetes.PodTemplate>
  kind: ConfigMap
  metadata:
    creationTimestamp: 2018-02-07T20:22:27Z
    labels:
      role: jenkins-slave
    name: jenkins-slave
    namespace: myproject
    resourceVersion: "2539"
    selfLink: /api/v1/namespaces/myproject/configmaps/jenkins-slave
    uid: 9dcdf7f6-0c44-11e8-998e-68f7287398a7
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

With perhaps multiple config maps.  If you can confirm the specifics, even provide the yaml of the config maps that existed when you saw this, that might help me to reproduce on my end and debug.

2) Do you have any of the jenkins pod logs when this occurred?  Again, the information there when the error occurs could shed some light for me.

3) If possible, try to reproduce with a logger for io.fabric8.jenkins.openshiftsync.ConfigMapWatcher created and set to log at all levels.  Then like 2), provide the jenkins pod logs and I can take a look.

thanks

**Comment by victorvoronin** [comment]
@gabemontero I am very sorry. At the moment I don't have time to create test cluster and describe my concern. I will do it a bit later.

**Comment by coreydaley** [comment]
@victorvoronin any updates on this?

**Comment by bszeti** [comment]
Hi, I run into a similar issue on an OpenShift cluster.
- config map is created with one PodTemplate
- when jenkins-slave label is added to the configmap Jenkins adds the template to k8s plugin config multiple (9!) times.
In Jenkins log these lines also appear multiple times:
Apr 17, 2018 3:08:24 AM io.fabric8.jenkins.openshiftsync.JenkinsUtils removePodTemplate
INFO: Removing PodTemplate: mymaven
Apr 17, 2018 3:08:24 AM io.fabric8.jenkins.openshiftsync.JenkinsUtils removePodTemplate
INFO: Removing PodTemplate: mymaven
...
Apr 17, 2018 3:08:24 AM io.fabric8.jenkins.openshiftsync.JenkinsUtils addPodTemplate
INFO: Adding PodTemplate: mymaven
Apr 17, 2018 3:08:24 AM io.fabric8.jenkins.openshiftsync.JenkinsUtils addPodTemplate
INFO: Adding PodTemplate: mymaven

Openshift cluster version: v3.6
openshift3/jenkins-2-rhel7: v3.9 -> Openshift Sync plugin 1.0.7

I tried to reproduce with CDK3.3, but it worked there correctly. I try to investigate more, but wanted to add a comment to keep the ticket open.

**Comment by gabemontero** [comment]
OK thanks @bszeti 

Is there any chance you can provide the config map yaml you used ... we can see if it differs from our attempts to reproduce this.

If you can investigate more, can you try to reproduce with  io.fabric8.jenkins.openshiftsync.ConfigMapWatcher created and set to log at all levels

I do have a theory on this was well ....I recently discovered that we may be leaking watchers when config changes are made from the jenkins console (possibly even if your are not changing the sync plugin config). 

I recently fixed this with commit ddb61b18e3ff84801ea0654d6ba5ad3938f7e9c3 and v1.0.10

If there are multiple config map watchers running, this could explain why there are multiple injections of the templates. 

The trace I noted above would help prove this, as could trying to reproduce with the later version (and not seeing the duplication).

**Comment by bszeti** [comment]
@gabemontero Thanks, I can confirm that the leaking watchers caused the podTemplate duplication. Every time a Kubernetes plugin modification was done via the UI another ConfigMapWatcher was started. The issue seems to be resolved with plugin v1.0.13.
